app/scripts/metamask-controller.js

```diff
@@ -4665,6 +4680,42 @@ export default class MetamaskController extends EventEmitter {
           },
         );
       }
+
+      ///: BEGIN:ONLY_INCLUDE_IF(multi-srp)
+      const client = new KeyringInternalSnapClient({
+        messenger: this.controllerMessenger,
+        snapId: SOLANA_WALLET_SNAP_ID,
+      });
+      // TODO: Use `withKeyring` instead.
+      const keyring = await this.getSnapKeyring();
+      for (let index = 0; ; index++) {
+        const discovered = await client.discoverAccounts(
+          [SolScope.Mainnet, SolScope.Testnet, SolScope.Devnet],
+          entropySource,
+          index,
+        );
+
+        // We stop discovering accounts if none got discovered for that index.
+        if (discovered.length === 0) {
+          break;
+        }
+
+        await Promise.allSettled(
+          discovered.map(async (discoveredAccount) => {
+            const options = {
+              derivationPath: discoveredAccount.derivationPath,
+              entropySource,
+            };
+
+            await keyring.createAccount(SOLANA_WALLET_SNAP_ID, options, {
+              displayConfirmation: false,
+              displayAccountNameSuggestion: false,
+              setSelectedAccount: false,
+            });
+          }),
+        );
```

Comment:
IDK what to think about using the `SnapKeyring` while by-passing the `KeyringController` global mutex in such context.

Since everything will happen asynchronously and that it will mutate the Snap keyring state too, we might want to use that mutex.

So actually, we should probably use `withKeyring` here.

We might need to call `await this.getSnapKeyring();` beforehand, so the Snap keyring gets constructed on the `KeyringController`. (I haven't tried with the option `createIfMissing`, that could also work with it?)